pulsar-broker/src/main/java/org/apache/pulsar/broker/service/ServerCnx.java

```diff
@@ -243,11 +244,18 @@ public ServerCnx(PulsarService pulsar) {
         this.maxPendingBytesPerThread = conf.getMaxMessagePublishBufferSizeInMB() * 1024L * 1024L
                 / conf.getNumIOThreads();
         this.resumeThresholdPendingBytesPerThread = this.maxPendingBytesPerThread / 2;
+        this.connectionController = new ConnectionController.DefaultConnectionController(conf);
     }
 
     @Override
     public void channelActive(ChannelHandlerContext ctx) throws Exception {
         super.channelActive(ctx);
+        if (!connectionController.increaseConnection(remoteAddress)) {
+            ctx.channel().writeAndFlush(Commands.newError(-1, ServerError.NotAllowedError
+                    , "Reached the maximum number of connections"));
```

Comment:
Users couldn't distinguish the specific error from this message, reach the maximum of a broker or reach the maximum of per IP.

Reply:
Done